weldx/util.py

```diff
@@ -1201,3 +1193,78 @@ def time_ref(self, value: pd.Timestamp):
                 self._obj.time.attrs["time_ref"] = value  # set new time_ref value
             else:
                 self._obj.time.attrs["time_ref"] = value
+
+
+class _Eq_compare_nested:
+    """Compares nested data structures like lists, sets, tuples, arrays, etc."""
+
+    compare_funcs = {
+        (np.ndarray, pint.Quantity): lambda x, y: np.all(x == y),
+        (xr.DataArray, xr.Dataset): lambda x, y: x.identical(y),
+    }
+
+    @staticmethod
+    def _compare(x, y):
+        if type(x) is not type(y):
+            return False
+        for types, func in _Eq_compare_nested.compare_funcs.items():
+            if isinstance(x, types):
+                return func(x, y)
+            return x == y
+
+    @staticmethod
+    def _visit(p, k, v, b):
+        """Traverses all elements in `compare_nested` argument a and b...
+
+        and tries to obtain the path `p` in `b` using boltons.iterutils.get_path.
+        The following cases can occur:
+        1. If the path does not exist in `b` a KeyError will be raised.
+        2. If the index `k` does not exist an IndexError is raised.
+        3. If the other path exists, a comparison will be made using `_compare`.
+           When the elements are not equal traversing `a` will be stopped
+           by raising a ValueError.
+        """
+        other_value = iterutils.get_path(b, p)[k]
+        if not _Eq_compare_nested._compare(v, other_value):
+            raise ValueError
+        return True
```

Comment:
```suggestion
        if not iterutils.default_enter(p, k, v)[1]:
            if not _Eq_compare_nested._compare(v, other_value):
                raise ValueError(f"{v=}\n{other_value=}")
        return True
```
We need to make sure not call the compare function on items that should be traversed deeper because we would end up calling `__eq__` on these as well (and failing)

try the following:
```python
a = {"l1": {"l2": np.arange(5)}}
b = {"l1": {"l2": np.arange(5)}}
assert ut.compare_nested(a, b)
```

The solution above is a little ugly but works...

Also the ValueError should report the compared instances for clarity

Reply:
I raise the value error just to break out of the iteration, not to make any useful error messages for the user, since the outer function just returns a boolean (I'd prefer such an interface for simplicity).